compiler/qsc_ast/src/ast.rs

```diff
@@ -1968,3 +1994,129 @@ impl ImportOrExportItem {
+pub struct TypeParameter {
+    /// Class constraints specified for this type parameter -- any type variable passed in
+    /// as an argument to these parameters must satisfy these constraints.
+    pub constraints: ClassConstraints,
+    /// The name of the type parameter.
+    pub ty: Ident,
+    /// The span of the full type parameter, including its name and its constraints.
+    pub span: Span,
+}
+
+impl WithSpan for TypeParameter {
+    fn with_span(self, span: Span) -> Self {
+        Self {
+            span,
+            ty: self.ty.with_span(span),
+            ..self
+        }
+    }
+}
+
+impl TypeParameter {
+    /// Instantiates a new `TypeParameter` with the given type name, constraints, and span.
+    #[must_use]
+    pub fn new(ty: Ident, bounds: ClassConstraints, span: Span) -> Self {
+        Self {
+            ty,
+            constraints: bounds,
+            span,
+        }
+    }
+}
+
+impl std::fmt::Debug for TypeParameter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // 'A: Eq + Ord + Clone
+        write!(
+            f,
+            "{}{}{:?}",
+            self.ty,
+            if self.constraints.0.is_empty() {
+                ""
+            } else {
+                ": "
+            },
+            self.constraints
+        )
+    }
+}
+
+/// A list of class constraints, used when constraining a type parameter.
+#[derive(Default, PartialEq, Eq, Clone, Hash)]
+pub struct ClassConstraints(pub Box<[ClassConstraint]>);
+
+/// An individual class constraint, used when constraining a type parameter.
+/// To understand this concept, think of parameters in a function signature -- the potential arguments that can
+/// be passed to them are constrained by what type is specified. Type-level parameters are no different, and
+/// the type variables that are passed to a type parameter must satisfy the constraints specified in the type parameter.
+#[derive(Default, PartialEq, Eq, Clone, Hash, Debug)]
+pub struct ClassConstraint {
+    /// The name of the constraint.
+    pub name: Ident,
+    /// Parameters for a constraint. For example, `Iterator` has a parameter `T` in `Iterator<T>` -- this
+    /// is the type of the item that is coming out of the iterator.
+    pub parameters: Box<[ConstraintParameter]>,
+}
+
+/// An individual constraint parameter is a type that is passed to a constraint, such as `T` in `Iterator<T>`.
+/// #[derive(Default, `PartialEq`, Eq, Clone, Hash, Debug)]
+#[derive(Default, PartialEq, Eq, Clone, Hash, Debug)]
+
+pub struct ConstraintParameter {
+    /// The type variable being passed as a constraint parameter.
+    pub ty: Ty,
+}
+
+impl WithSpan for ConstraintParameter {
+    fn with_span(self, span: Span) -> Self {
+        Self {
+            ty: self.ty.with_span(span),
+        }
+    }
+}
+
+impl ConstraintParameter {
+    /// Getter for the `ty` field.
+    #[must_use]
+    pub fn ty(&self) -> &Ty {
+        &self.ty
+    }
```

Comment:
Why these trivial getters for public fields? Again I'd argue for consistency across the AST nodes - the existing convention seems to be accessing fields directly.